taichi/backends/metal/kernel_manager.cpp

```diff
@@ -415,8 +483,22 @@ class HostMetalCtxBlitter {
              ActionArg("offset_in_bytes", (int64)arg.offset_in_mem)});
       }
       if (arg.is_array) {
-        const void *host_ptr = host_ctx_->get_arg<void *>(i);
-        std::memcpy(device_ptr, host_ptr, arg.stride);
+        if (host_ctx_->is_device_allocations[i]) {
+          // There is no way to write from host into Ndarray directly (yet), so
+          // we don't have to do anything here.
+        } else {
+          const void *host_ptr = host_ctx_->get_arg<void *>(i);
+          const auto alloc_n_sz = cti_kernel_->ext_arr_arg_to_dev_alloc.at(i);
+          auto buf_mem = alloc_mapper_->find(alloc_n_sz.alloc);
+          TI_ASSERT(buf_mem.buffer != nullptr);
+          auto *mem = buf_mem.mem;
+          TI_ASSERT(mem != nullptr);
+          // NOTE: DO NOT use `alloc_n_sz.size`. That is the size for the
+          // allocation, NOT the array size.
+          const auto arr_size = host_ctx_->array_runtime_sizes[i];
+          std::memcpy(mem->ptr(), host_ptr, arr_size);
```

Comment:
We already have a DeviceAllocation for external arrays in `ext_arr_arg_to_dev_alloc`, if that's what you mean?

Reply:
@k-ye ohh now I see they are equivalent as mapping and unmapping a DeviceAllocation. I was mainly wondering why we used `find()` to find the pointer directly instead of mapping the DeviceAllocation from `ext_arr_arg_to_dev_alloc` with Device.